File: SexyBiscuit.Engine/Core/MissingComponent.cs
using System.Text.Json;

namespace SexyBiscuit.Engine.Core;

/// <summary>
/// Stands in for a component whose type could not be resolved when a scene loaded — a game
/// class that failed to compile, was renamed, or comes from an assembly that is not loaded.
/// </summary>
/// <remarks>
/// The loader used to drop such components silently, so one bad build turned every instance
/// of a project component into lost data the moment the scene was saved again. The placeholder
/// keeps the original type name and property JSON and writes them back unchanged, the way
/// Unity's "missing script" does. It does nothing at runtime.
/// </remarks>
public sealed class MissingComponent : Component
{
    /// <summary>The type name exactly as the scene file wrote it.</summary>
    public string TypeName { get; set; } = "";

    /// <summary>The property bag exactly as the scene file wrote it.</summary>
    public Dictionary<string, JsonElement>? Properties { get; set; }
}

/// <summary>
/// Remembers the actor class a scene file asked for when that class could not be resolved.
/// The actor loads as a plain <see cref="Actor"/> with all of its components, and the class
/// name is written back on save so nothing is lost.
/// </summary>
public sealed class MissingActorClass : Component
{
    public string ClassName { get; set; } = "";
}

File: html5/src/core/MissingComponent.js
// -----------------------------------------------------------------------------
// MissingComponent / MissingActorClass — placeholders that stop a scene losing
// data when a type cannot be resolved.
//
// This matters far more in the browser than it does in C#. A shared project's
// behaviour may be written in C# and compiled into an assembly no web runtime
// can load; those component types will always be unresolvable here. Keeping the
// type name and its property JSON verbatim means the HTML5 editor can open, edit
// and save such a scene, and the C# engine still finds its components intact.
// -----------------------------------------------------------------------------

import { Component } from './Component.js';
import { registerComponent } from './TypeRegistry.js';

/** Holds the data of a component whose type could not be resolved. Does nothing. */
export class MissingComponent extends Component {
    static schema = {};

    constructor() {
        super();
        /** The `type` string exactly as the file had it. */
        this.typeName = '';
        /** The raw property bag, written back untouched on save. */
        this.properties = null;
    }

    toString() { return `MissingComponent(${this.typeName})`; }
}
registerComponent(MissingComponent, { category: 'Core', hidden: true });

/** Marks an actor whose `class` could not be resolved, so the name survives a save. */
export class MissingActorClass extends Component {
    static schema = {};

    constructor() {
        super();
        this.className = '';
        /** The subclass's own property bag, written back untouched. */
        this.properties = null;
    }

    toString() { return `MissingActorClass(${this.className})`; }
}
registerComponent(MissingActorClass, { category: 'Core', hidden: true });
